sorts/tim_sort.ts

```diff
@@ -87,20 +96,93 @@ const merge = <T>(
  * @param compare The comparator function defining the order of elements.
  */
 export const timSort = <T>(arr: T[], compare: Comparator<T>): void => {
-  const n = arr.length
+  const length = arr.length
+
+  /**
+   * Reverses a portion of the array.
+   *
+   * @param start The starting index of the portion to reverse.
+   * @param end The ending index of the portion to reverse.
+   */
+  const reverseRange = (start: number, end: number): void => {
+    while (start < end) {
+      const temp = arr[start]
+      arr[start++] = arr[end]
+      arr[end--] = temp
+    }
+  }
+
+  /**
+   * Identifies runs and sorts them using insertion sort.
+   *
+   * @param start The starting index of the range to find runs.
+   * @param end The ending index of the range to find runs.
+   */
+  const findRunsAndSort = (start: number, end: number): void => {
+    for (let currIdx = start + 1; currIdx <= end; currIdx++) {
+      const currentElement = arr[currIdx]
+      let prevIdx = currIdx - 1
 
-  for (let i = 0; i < n; i += MIN_MERGE) {
-    insertionSort(arr, i, Math.min(i + MIN_MERGE - 1, n - 1), compare)
+      while (prevIdx >= start && compare(arr[prevIdx], currentElement) > 0) {
+        arr[prevIdx + 1] = arr[prevIdx]
+        prevIdx--
+      }
+      arr[prevIdx + 1] = currentElement
+    }
   }
 
-  for (let size = MIN_MERGE; size < n; size *= 2) {
-    for (let left = 0; left < n; left += 2 * size) {
-      const mid = left + size - 1
-      const right = Math.min(left + 2 * size - 1, n - 1)
+  /**
+   * Merges runs in the array.
+   *
+   * @param minRunLength The minimum length of a run.
+   */
+  const mergeRuns = (minRunLength: number): void => {
```

Comment:
This is still Theta(n log n)?

Reply:
Sorry for missing there. I will update the implementation.

Reply:
Please explain why you believe that this is asymptotically more efficient in the best case.

As far as I can see, the outer loop is still logarithmic since minRunLength is still roughly constant. The inner loop is still linear time. This would still give linearithmic time.

Reply:
I forgot that. As you said in the previous comments, I have been implementing a more efficient `mergeRuns` function.

Reply:
Do you have any suggestions?

Reply:
You're basically doing a merge sort with fixed size subarrays here. This will always have Theta(n log n) runtime.

What you should be doing instead is pulling adjacent sorted runs from a stack and merging them until you're done.

As a reference, the [Wikipedia article](https://en.wikipedia.org/wiki/Timsort) looks decent. You can probably also find good references in the source section of said Wikipedia article.